oprf/group/group.go

```diff
@@ -0,0 +1,117 @@
+package group
+
+import (
+	"hash"
+	"math/big"
+)
+
+// CiphersuiteID is
+type CiphersuiteID int
+
+const (
+	// P256 is the group generated by P-256 elliptic curve.
+	P256 CiphersuiteID = iota
+)
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+// Ciphersuite object determines the prime-order group (pog) that is used for
+// performing the OPRF operations, along with the different hash function
+// definitions.
+// Should be created using New, using a CiphersuiteID constant.
+type Ciphersuite struct {
+	// name of the ciphersuite
+	name string
+
+	// context
+	ctx string
+
+	// Suite instantiation for performing the OPRF operations.
+	suite Suite
+}
+
+// Suite is an interface that defines operations within additive
+// groups of prime order. This is the setting in which the OPRF operations
+// take place.
+type Suite interface {
+	// Creates a new Suite object
+	New(CiphersuiteID) (Suite, error)
+
+	// Returns the identifying name of the group
+	Name() string
+
+	// Returns the canonical (fixed) generator for defined group
+	Generator() *Point
+
+	// Returns the order of the canonical generator in the group.
+	Order() *big.Int
+
+	// Base hash function used in conjunction with the Ciphersuite
+	Hash() hash.Hash
+
+	// Returns k * G, where G is the canonical generator of the group, and k is
+	// some scalar value provided as input.
+	ScalarMult(*Scalar) (*Point, error)
+
+	// Performs a transformation to encode bytes as a Point object in the
+	// group.
+	HashToGroup([]byte) (*Point, error)
+
+	// Performs a transformation to encode bytes as a Scalar object in the
+	// appropiate group.
+	HashToScalar([]byte) (Scalar, error)
+
+	// Samples a random scalar value from the field of scalars defined by the
+	// group order.
+	RandomScalar() (*Scalar, error)
+
+	// Returns the ByteLength of objects associated with the Ciphersuite
+	ByteLength() int
+}
+
+// Point is the interface that represents points in a given
+// Ciphersuite instantiation.
+type Point interface {
+	// New constructs a Point object for the associated Suite
+	// instantiation
+	New(Suite) *Point
+
+	// Returns a bool indicating that the Point is valid
+	IsValid() bool
```

Comment:
I wonder if it makes sense to have this function in the first place. Can we just ensure that all points are valid? When would we need an invalid point?

Reply:
Might be. I guess that at some point you might receive a public key (which is a Point) from a Client, for example, that generated it not by using our library but some other library. As it is not possible at that point to know if the public key was correctly generated, this function can be used.

Reply:
Oh, my point was that we can simply fail deserializing such points, rather than deserialize them and then require applications to check "IsValid". Would that work?